machineconfiguration/v1alpha1/types_machineosbuild.go

```diff
@@ -0,0 +1,218 @@
+	ExitReason string `json:"exitReason"`
+}
+
+// BuoldConfigMapObjectReference refers to the name and contents of a configmap used in a previous build
+type BuildConfigMapObjectReference struct {
+	// name of the configmap
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+	// Data is the string data in the configmap
+	// +optional
+	Data ConfigMapData `json:"data"`
+	// Binary data is the binary data in the configmap
+	// +optional
+	BinaryData ConfigMapData `json:"binaryData"`
+}
+
+// ConfigMapData is used to map the data in the OCB configmaps so that they can be viewed in the object's statuss
+type ConfigMapData struct {
+	// key in the configmap pointing to the specific data
+	// +kubebuilder:validation:Required
+	Key string `json:"key"`
+	// value is the data stored in the configmap
+	// +optional
+	Value string `json:"value"`
+	// binaryValue is the data stored in the configmap
+	// +optional
+	BinaryValue []byte `json:"binaryValue"`
+}
+
+// Refers to the name of a MachineConfigPool (e.g., "worker", "infra", etc.):
+type MachineConfigPoolReference struct {
+	// name of the MachineConfigPool object.
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+}
+
+// Refers to the name of a rendered MachineConfig (e.g., "rendered-worker-ec40d2965ff81bce7cd7a7e82a680739", etc.):
+type RenderedMachineConfigReference struct {
+	// name is the name of the rendered MachineConfig object.
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+}
+
+// Refers to the name of a (future) MachineOSImage (e.g., "worker-os-image-167651b10ec98af17971d6a47df9e22f", etc.):
+type MachineOSImageReference struct {
+	// name is the name of the MachineOSImage object attached to this MachineOSBuild.
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+}
+
+// Refers to the name of an image registry push/pull secret needed in the build process.
+type ImageSecretObjectReference struct {
+	// name is the name of the secret used to push or pull this MachineOSBuild object.
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+}
+
+// BuildInputs holds all of the information needed to trigger a build
+type BuildInputs struct {
+	// imageBuilderType specifies the backend to be used to build the image.
+	// Valid options are: OpenShiftImageBuilder, PodImageBuilder, and Default (OpenShiftImageBuilder)
+	ImageBuilderType MachineOSImageBuilderType `json:"imageBuilderType"`
+	// baseOSExtensionsImageURL is the base Extensions image used in the build process
+	// +kubebuilder:validation:Required
+	BaseOSExtensionsImageURL string `json:"baseOSExtensionsImageURL"`
+	// baseOSImageURL is the base OSImage we use to build our custom image.
+	// +kubebuilder:validation:Required
+	BaseOSImageURL string `json:"baseOSImageURL"`
+	// baseImagePullSecret is the secret used to pull the base image.
+	// +kubebuilder :validation:Required
+	BaseImagePullSecret ImageSecretObjectReference `json:"baseImagePullSecret"`
```

Comment:
we talked about this in the review session but there really only is a single pull secret for the cluster so I am not sure what this supposed to be for.  Is this a path to a pullsecret on disk?

Reply:
The reason we have this is because as of 4.16, the user will need to specify their own base image pull secret, we do not have the plumbing to get this for them.

The plan in 4.17+ is to add some sort of validation so they cannot edit this? Though @cheesesashimi can you think of a scenario where we want to allow them to specify a custom base image pull secret?

Reply:
I could see them potentially asking for that in a Hypershift context. @yuqi-zhang In Hypershift clusters, is it common for the guest clusters to use the same global pull secret as the control-plane cluster? Is there a situation where it would be desirable for someone to do that?